File: go/evaluate.go
package likeness

import (
	"strconv"
	"strings"
	"time"
)

const dayMs = int64(86400000)

func durationMs(v *Value) int64 {
	switch v.Unit {
	case "d":
		return int64(v.N) * dayMs
	case "w":
		return int64(v.N) * 7 * dayMs
	case "mo":
		return int64(v.N) * 30 * dayMs
	case "y":
		return int64(v.N) * 365 * dayMs
	}
	return 0
}

func parseRfc3339(s string) (int64, bool) {
	t, err := time.Parse(time.RFC3339, s)
	if nil != err {
		return 0, false
	}
	return t.UnixMilli(), true
}

func resolveDate(v *Value, clockMs int64) (int64, bool) {
	if "date" == v.K {
		day := v.V
		switch v.V {
		case "today":
			day = time.UnixMilli(clockMs).UTC().Format("2006-01-02")
		case "yesterday":
			day = time.UnixMilli(clockMs - dayMs).UTC().Format("2006-01-02")
		}
		return parseRfc3339(day + "T00:00:00Z")
	}
	if "dur" == v.K {
		return clockMs - durationMs(v), true
	}
	return 0, false
}

func fieldValue(n Note, field string) (any, bool) {
	switch field {
	case "tag":
		if nil == n.Tags {
			return nil, false
		}
		return n.Tags, true
	case "space":
		if nil == n.Space {
			return nil, false
		}
		return *n.Space, true
	case "source":
		return n.Source, true
	case "instance":
		return n.Instance, true
	case "status":
		if nil == n.Status {
			return nil, false
		}
		return *n.Status, true
	case "state":
		if nil == n.State {
			return nil, false
		}
		return *n.State, true
	case "title":
		return n.Title, true
	case "body":
		if nil == n.Body {
			return nil, false
		}
		return *n.Body, true
	case "created":
		return n.Created, true
	case "updated":
		return n.Updated, true
	}
	return nil, false
}

func asText(v *Value) string {
	if "dur" == v.K {
		return strconv.Itoa(v.N) + v.Unit
	}
	return v.V
}

// Evaluate answers whether a note matches a selector, given the injected clock.
func Evaluate(node *Node, note Note, clockMs int64) bool {
	switch node.T {
	case "or":
		for _, k := range node.Kids {
			if Evaluate(k, note, clockMs) {
				return true
			}
		}
		return false
	case "and":
		for _, k := range node.Kids {
			if !Evaluate(k, note, clockMs) {
				return false
			}
		}
		return true
	case "not":
		return !Evaluate(node.Kid, note, clockMs)
	case "text":
		// A bareword with no field is full text. With no source-side search it
		// is a case-insensitive substring over title and body, and the
		// capability matrix says so rather than implying a relevance score
		// nobody computed.
		q := strings.ToLower(node.V)
		if strings.Contains(strings.ToLower(note.Title), q) {
			return true
		}
		return nil != note.Body && strings.Contains(strings.ToLower(*note.Body), q)
	}

	have, present := fieldValue(note, node.Field)
	want := node.Value

	for _, o := range OrderingOps {
		if o != node.Op {
			continue
		}
		s, ok := have.(string)
		if !present || !ok {
			return false
		}
		haveMs, ok := parseRfc3339(s)
		if !ok {
			return false
		}
		wantMs, ok := resolveDate(want, clockMs)
		if !ok {
			return false
		}
		switch node.Op {
		case ">":
			return wantMs < haveMs
		case ">=":
			return wantMs <= haveMs
		case "<":
			return haveMs < wantMs
		case "<=":
			return haveMs <= wantMs
		}
	}

	text := strings.ToLower(asText(want))
	eq := func(h any) bool {
		if !present || nil == h {
			return false
		}
		s, ok := h.(string)
		return ok && strings.ToLower(s) == text
	}

	if list, ok := have.([]string); present && ok {
		hit := false
		for _, x := range list {
			if strings.ToLower(x) == text {
				hit = true
				break
			}
		}
		if "!=" == node.Op {
			return !hit
		}
		return hit
	}
	if "=" == node.Op {
		return eq(have)
	}
	if "!=" == node.Op {
		return !eq(have)
	}
	// ':' is contains for free-text fields and equality for the rest, which is
	// what people mean by `title:retention` and by `status:done`.
	if "title" == node.Field || "body" == node.Field {
		if !present {
			return false
		}
		s, ok := have.(string)
		return ok && strings.Contains(strings.ToLower(s), text)
	}
	return eq(have)
}

File: typescript/src/evaluate.ts
/* Evaluate a parsed selector against a projected note, locally.
 *
 * Structured terms are evaluated IDENTICALLY EVERYWHERE: pushed down to the
 * source where the capability matrix says it can filter on them, and applied
 * here where it cannot - same answer either way, which is a corpus case rather
 * than a claim. For Stage 1 no source can filter, so everything lands here.
 *
 * Durations are relative to the INJECTED clock, never to the wall clock. An
 * entry that asserted on `now` would pass for one second and fail forever
 * after.
 */

import type { Node, Value } from './selector.js'
import type { Note } from './note.js'

const DAY = 86400000

function durationMs (v: Extract<Value, { k: 'dur' }>): number {
  switch (v.unit) {
    case 'd': return v.n * DAY
    case 'w': return v.n * 7 * DAY
    // A month is 30 days and a year is 365. Calendar arithmetic differs between
    // language standard libraries in ways nobody agrees on, and a selector
    // window is a rough question. Fixed multipliers are pinned in the corpus.
    case 'mo': return v.n * 30 * DAY
    case 'y': return v.n * 365 * DAY
  }
}

function resolveDate (v: Value, clockMs: number): number | null {
  if (v.k === 'date') {
    if (v.v === 'today') return Date.parse(new Date(clockMs).toISOString().slice(0, 10) + 'T00:00:00Z')
    if (v.v === 'yesterday') return Date.parse(new Date(clockMs - DAY).toISOString().slice(0, 10) + 'T00:00:00Z')
    return Date.parse(v.v + 'T00:00:00Z')
  }
  if (v.k === 'dur') return clockMs - durationMs(v)
  return null
}

function fieldValue (n: Note, field: string): unknown {
  switch (field) {
    case 'tag': return n.tags
    case 'space': return n.space
    case 'source': return n.source
    case 'instance': return n.instance
    case 'status': return n.status
    case 'state': return n.state
    case 'title': return n.title
    case 'body': return n.body
    case 'created': return n.created
    case 'updated': return n.updated
    default: return undefined
  }
}

function asText (v: Value): string {
  if (v.k === 'word') return v.v
  if (v.k === 'date') return v.v
  return String(v.n) + v.unit
}

export function evaluate (node: Node, note: Note, clockMs: number): boolean {
  switch (node.t) {
    case 'or': return node.kids.some(k => evaluate(k, note, clockMs))
    case 'and': return node.kids.every(k => evaluate(k, note, clockMs))
    case 'not': return !evaluate(node.kid, note, clockMs)
    case 'text': {
      // A bareword with no field is full text. With no source-side search it is
      // a case-insensitive substring over title and body, and the capability
      // matrix says so rather than implying a relevance score nobody computed.
      const q = node.v.toLowerCase()
      return note.title.toLowerCase().includes(q) ||
        (undefined !== note.body && note.body.toLowerCase().includes(q))
    }
    case 'cmp': {
      const have = fieldValue(note, node.field)
      const want = node.value

      if (node.op === '>' || node.op === '<' || node.op === '>=' || node.op === '<=') {
        if ('string' !== typeof have) return false
        const haveMs = Date.parse(have)
        const wantMs = resolveDate(want, clockMs)
        if (null === wantMs || Number.isNaN(haveMs)) return false
        // `updated>14d` reads as "changed within the last 14 days", so a
        // duration on the right of `>` is a LOWER BOUND ON THE TIMESTAMP.
        switch (node.op) {
          case '>': return haveMs > wantMs
          case '>=': return haveMs >= wantMs
          case '<': return haveMs < wantMs
          case '<=': return haveMs <= wantMs
        }
      }

      const text = asText(want)
      const eq = (h: unknown) => {
        if (undefined === h || null === h) return false
        return String(h).toLowerCase() === text.toLowerCase()
      }
      const contains = (h: unknown) => {
        if (undefined === h || null === h) return false
        return String(h).toLowerCase().includes(text.toLowerCase())
      }

      if (Array.isArray(have)) {
        const hit = have.some(x => String(x).toLowerCase() === text.toLowerCase())
        return node.op === '!=' ? !hit : hit
      }
      if (node.op === '=') return eq(have)
      if (node.op === '!=') return !eq(have)
      // ':' is contains for free-text fields and equality for the rest, which
      // is what people mean by `title:retention` and by `status:done`.
      if (node.field === 'title' || node.field === 'body') return contains(have)
      return eq(have)
    }
  }
}
